dali/base/dautils.cpp

```diff
@@ -3083,3 +3113,262 @@ bool traceAllTransactions(bool on)
+        IMPLEMENT_IINTERFACE;
+        MemoryBuffer mb;
+        unsigned lockCount;
+
+        bool first()
+        {
+            mb.reset();
+            mb.read(lockCount);
+
+            return next();
+        }
+
+        bool next()
+        {
+            cur.clear();
+            if (mb.getPos()>=mb.length())
+                return false;
+            cur.setown(deserializeLockData(mb));
+            return true;
+        }
+
+        bool isValid()
+        {
+            return cur.get()!=NULL;
+        }
+
+        IPropertyTree  & query()
+        {
+            return *cur;
+        }
+    } *fai = new CLockDataIterator;
+    mb.swapWith(fai->mb);
+    return fai;
+}
+
+IPropertyTreeIterator *CLockDataHelper::getLockDataTreeIterator(MemoryBuffer &lockInfo)
+{
+    return deserializeLockDataIterator(lockInfo);
+}
+
+StringBuffer &CLockDataHelper::formatLockData(MemoryBuffer &lockInfo, StringBuffer &out)
+{
+    StringAttr xPath;
+    lockInfo.read(xPath);
+    out.append("Locks on path: /").append(xPath).newline();
+    out.append("Endpoint            |SessionId       |ConnectionId    |mode    |time(duration)]").newline().newline();
+
+    unsigned lockCount;
+    lockInfo.read(lockCount);
+    if (lockCount < 1)
+    {
+        out.newline();
+        return out;
+    }
+
+    unsigned msNow = msTick();
+    CDateTime time;
+    time.setNow();
+
+    for (unsigned l = 0; l < lockCount; l++)
+    {
+        StringAttr ep;
+        __int64 connId, sessId;
+        unsigned mode, timeLockObtained;
+        lockInfo.read(ep);
+        lockInfo.read(sessId);
+        lockInfo.read(connId);
+        lockInfo.read(mode);
+        lockInfo.read(timeLockObtained);
+
+        CDateTime timeLocked;
+        unsigned lockedFor = msNow-timeLockObtained;
+        time_t tt = time.getSimple() - (lockedFor/1000);
+        timeLocked.set(tt);
+        StringBuffer timeStr;
+        timeLocked.getString(timeStr);
+        out.appendf("%-20s|%-16" I64F "x|%-16" I64F "x|%-8x|%s(%d ms)", ep.str(), sessId, connId, mode,
+            timeStr.str(), lockedFor).newline();
+    }
+    return out;
+}
+
+bool CLockDataHelper::checkEP(const char *ep, const char *ipPattern)
+{
+    StringBuffer ips;
+    while (*ep && (*ep != ':'))
+        ips.append(*(ep++));
+    return WildMatch(ips.str(), ipPattern);
+}
+
+void CLockDataHelper::formatLock(IPropertyTree &query, bool fileLock, unsigned formatType, bool &headerDone, const char *ipPattern,
+    const char *xpathPattern, Int64Array *connIds, StringBuffer &out)
+{
+    unsigned epCount =  query.getPropInt(getLockDataFieldName(LDFEPCount));
+    if (epCount < 1)
+        return;
+
+    StringBuffer xpath, file;
+    query.getProp(getLockDataFieldName(LDFXPath), xpath);
+
+    CDfsLogicalFileName lfn;
+    lfn.xPathToLName(xpath.insert(0, '/').str(), file);
+
+    if ((file.length() > 0) != fileLock)
+        return;
+    if (xpathPattern && (!xpath.length() || !WildMatch(xpath.str(), xpathPattern))
```

Comment:
Doesn't look like !xpath.length() can ever be true, because of the insert above.